packages/components/src/box-model-overlay/index.tsx

```diff
@@ -0,0 +1,271 @@
+			showValues.padding?.right ? paddingRight : '0'
+		);
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-padding-bottom',
+			showValues.padding?.bottom ? paddingBottom : '0'
+		);
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-padding-left',
+			showValues.padding?.left ? paddingLeft : '0'
+		);
+	}, [ targetRef, showValues.margin, showValues.padding ] );
+
+	// Make the imperative `update` method available via `ref`.
+	useImperativeHandle( ref, () => ( { update } ), [ update ] );
+
+	const getAnchorRect = useCallback(
+		() => targetRef.current.getBoundingClientRect(),
+		[ targetRef ]
+	);
+
+	// Completely skip rendering the popover if none of showValues is true.
+	const shouldShowOverlay = useMemo(
+		() =>
+			Object.values( showValues.margin ?? {} ).some(
+				( value ) => value === true
+			) ||
+			Object.values( showValues.padding ?? {} ).some(
+				( value ) => value === true
+			),
+		[ showValues.margin, showValues.padding ]
+	);
+
+	useEffect( () => {
+		const target = targetRef.current;
+
+		if ( ! shouldShowOverlay || ! target ) {
+			return;
+		}
+
+		const defaultView = target.ownerDocument.defaultView;
+
+		update();
+
+		const resizeObserver = new defaultView.ResizeObserver( update );
+		const mutationObserver = new defaultView.MutationObserver( update );
+
+		// Observing size changes.
+		resizeObserver.observe( target, { box: 'border-box' } );
+		// Observing padding and margin changes.
+		mutationObserver.observe( target, {
+			attributes: true,
+			attributeFilter: [ 'style' ],
+		} );
+
+		return () => {
+			resizeObserver.disconnect();
+			mutationObserver.disconnect();
+		};
+	}, [ targetRef, shouldShowOverlay, update ] );
+
+	return shouldShowOverlay ? (
+		<OverlayPopover
+			position="middle center"
+			__unstableForcePosition
+			__unstableForceXAlignment
+			getAnchorRect={ getAnchorRect }
+			animate={ false }
+			aria-hidden="true"
+			focusOnMount={ false }
+			ref={ overlayRef }
+			{ ...props }
+		/>
+	) : null;
+} );
+
+const BoxModelOverlayWithChildren = forwardRef<
+	BoxModelOverlayHandle,
+	BoxModelOverlayPropsWithChildren
+>( ( { children, ...props }, ref ) => {
+	const targetRef = useRef< HTMLElement >();
+
+	return (
+		<>
+			{ cloneElement( Children.only( children ), { ref: targetRef } ) }
+			<BoxModelOverlayWithRef
+				{ ...props }
+				targetRef={ targetRef }
+				ref={ ref }
+			/>
+		</>
+	);
+} );
+
+const hasChildren = (
+	props: BoxModelOverlayProps
+): props is BoxModelOverlayPropsWithChildren => 'children' in props;
+
+const BoxModelOverlay = forwardRef<
+	BoxModelOverlayHandle,
+	BoxModelOverlayProps
+>( ( props, ref ) => {
+	if ( hasChildren( props ) ) {
+		return <BoxModelOverlayWithChildren { ...props } ref={ ref } />;
+	}
+
+	return <BoxModelOverlayWithRef { ...props } ref={ ref } />;
+} );
+
+export {
+	BoxModelOverlayProps,
+	BoxModelOverlayPropsWithChildren,
+	BoxModelOverlayPropsWithTargetRef,
+	BoxModelOverlayHandle,
+};
```

Comment:
These are only exported for the tests for now. Not sure how we can export them to the package level when we're still using `.js` in our entry file.